File: 백준/Gold/10026. 적록색약/적록색약.py
def bfs(ny,nx,key):
    q = deque()
    q.append((ny,nx))
    while q:
        y,x = q.popleft()
        for i in range(4):
            dy = diy[i] + y
            dx = dix[i] + x
            if dy<0 or dy>=N or dx<0 or dx>=N:
                continue
            if used[dy][dx] == 1:
                continue
            if arr[dy][dx] == key:
                used[dy][dx] = 1
                q.append((dy,dx))

def bfs_b(ny,nx,key):
    ky = 'X'
    if key == 'R':
        ky = 'G'
    if key == 'G':
        ky = 'R'
    q = deque()
    q.append((ny,nx))
    while q:
        y,x = q.popleft()
        for i in range(4):
            dy = diy[i] + y
            dx = dix[i] + x
            if dy<0 or dy>=N or dx<0 or dx>=N:
                continue
            if used_b[dy][dx] == 1:
                continue
            if arr[dy][dx] == key or arr[dy][dx]==ky:
                used_b[dy][dx] = 1
                q.append((dy,dx))


from collections import deque
N = int(input())
arr = [list(input()) for _ in range(N)]
diy = [0,0,1,-1]
dix = [1,-1,0,0]
used = [[0]*N for _ in range(N)]
used_b = [[0]*N for _ in range(N)]

a,b=0,0
b_list = []
for i in range(N):
    for j in range(N):
        if used[i][j] == 0:
            used[i][j] = 1
            bfs(i,j,arr[i][j])
            a += 1

        if used_b[i][j] == 0:
            used_b[i][j] = 1
            bfs_b(i,j,arr[i][j])
            b += 1

print(a,b)

File: 백준/Gold/10026. 적록색약/적록색약.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    static int N;
    static char[][] arr;
    static int[] diy = {0, 0, 1, -1};
    static int[] dix = {1, -1, 0, 0};
    static int[][] used;
    static int[][] used_b;

    public static void bfs(int ny, int nx, char key) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{ny, nx});
        used[ny][nx] = 1;

        while (!q.isEmpty()) {
            int[] current = q.poll();
            int y = current[0];
            int x = current[1];

            for (int i = 0; i < 4; i++) {
                int dy = diy[i] + y;
                int dx = dix[i] + x;

                if (dy < 0 || dy >= N || dx < 0 || dx >= N) {
                    continue;
                }
                if (used[dy][dx] == 1) {
                    continue;
                }
                if (arr[dy][dx] == key) {
                    used[dy][dx] = 1;
                    q.add(new int[]{dy, dx});
                }
            }
        }
    }

    public static void bfs_b(int ny, int nx, char key) {
        char ky = 'X';
        if (key == 'R') {
            ky = 'G';
        }
        if (key == 'G') {
            ky = 'R';
        }

        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{ny, nx});
        used_b[ny][nx] = 1;

        while (!q.isEmpty()) {
            int[] current = q.poll();
            int y = current[0];
            int x = current[1];

            for (int i = 0; i < 4; i++) {
                int dy = diy[i] + y;
                int dx = dix[i] + x;

                if (dy < 0 || dy >= N || dx < 0 || dx >= N) {
                    continue;
                }
                if (used_b[dy][dx] == 1) {
                    continue;
                }
                if (arr[dy][dx] == key || arr[dy][dx] == ky) {
                    used_b[dy][dx] = 1;
                    q.add(new int[]{dy, dx});
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        arr = new char[N][N];
        for (int i = 0; i < N; i++) {
            String line = sc.next();
            for (int j = 0; j < N; j++) {
                arr[i][j] = line.charAt(j);
            }
        }
        sc.close();

        used = new int[N][N];
        used_b = new int[N][N];

        int a = 0;
        int b = 0;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (used[i][j] == 0) {
                    bfs(i, j, arr[i][j]);
                    a++;
                }

                if (used_b[i][j] == 0) {
                    bfs_b(i, j, arr[i][j]);
                    b++;
                }
            }
        }

        System.out.println(a + " " + b);
    }
}